### PR TITLE
[ML] Reduce differences in regression and classification results between target platforms

### PR DESCRIPTION
**Pull request by tveasey**
This addresses some of the underlying causes for variation in regression and classification results for different platforms: `std::log` and `std::exp` don't always agree in the last decimal place and our split order predicate wasn't a total order.

### CODE REVIEW

File: docs/CHANGELOG.asciidoc [hunk 1]
```diff
@@ -47,6 +47,8 @@
 * Improve robustness of anomaly detection to bad input data. (See {ml-pull}1114[#1114].)
 * Adds new `num_matches` and `preferred_to_categories` fields to category output.
   (See {ml-pull}1062[#1062])
+* Ensure classification and regression results agree for all our target operating systems.
+  (See {ml-pull}1127[1127].)
```
Comment (droberts195) [inline docs/CHANGELOG.asciidoc:51]: The `1127` inside the square brackets should be preceded by `#`.

File: include/maths/CTools.h [hunk 2]
```diff
@@ -668,6 +668,17 @@ class MATHS_EXPORT CTools : private core::CNonInstantiatable {
     //! Compute \f$x^2\f$.
     static double pow2(double x) { return x * x; }
 
+    //! Compute a value from \p x which will be stable across platforms.
+    static double stable(double x) {
+        return STABLE_EPS * std::floor(x / STABLE_EPS + 0.5);
```
Comment (droberts195) [inline include/maths/CTools.h:673]: Could this cause problems if `x` is very large?  This is dividing by a number of the order of 1e-14 right?  So if `x` was 1e300, would anything go wrong?

Just to prove there are no nasty edge cases it would be good to add a unit test that exercises this function for these arguments:

* -1e305
* -1.0
* -1e-305
* 0.0
* 1e-305
* 1.0
* 1e305

Comment (tveasey) [reply include/maths/CTools.h:673]: This is a good point. It won't be near overflowing given how it's used, but, in fact, we have some tools for dealing with this correctly already I should have used.

Comment (tveasey) [reply include/maths/CTools.h]: Ok I just completely reworked this to drop the last bit in a way which doesn't rely on any standard library functions: using `frexp` and `ldexp` defeats the purpose. This implementation clearly doesn't overflow, so instead I added some tests for bitwise comparison of `stableLog` and `stableExp` for some random inputs. These would fail if they're different on any of our target platforms. See [7803c5c](https://github.com/elastic/ml-cpp/pull/1127/commits/7803c5c31fabb352f764b76b99885ad6f14382a2).

File: docs/CHANGELOG.asciidoc (hunk 1 above)
Comment (tveasey) [reply docs/CHANGELOG.asciidoc]: See [5f6500b](https://github.com/elastic/ml-cpp/pull/1127/commits/5f6500b278ec5c3b95dc83a25c51e9fb71231618).

File: include/maths/CTools.h [hunk 3]
```diff
@@ -22,6 +22,7 @@
 #include <boost/static_assert.hpp>
 
 #include <array>
+#include <cinttypes>
```
Comment (droberts195) [inline include/maths/CTools.h:25]: Maybe cstdint would be better - see https://stackoverflow.com/a/7597070/3790930

Comment (tveasey) [reply include/maths/CTools.h]: See [5d99c3e](https://github.com/elastic/ml-cpp/pull/1127/commits/5d99c3eec3b8d38f4d6ec429346cfebecf62f416).
